src/base/static/components/input-form/input-form.js

```diff
@@ -0,0 +1,833 @@
+    this.richTextImages = [];
+    this.coverImages = [];
+    this.visibleCategories = [];
+
+    this.validationExclusions = new Set([
+      constants.COMMON_FORM_ELEMENT_TYPENAME,
+      constants.MAP_DRAWING_TOOLBAR_TYPENAME,
+      constants.PUBLISH_CONTROL_TOOLBAR_TYPENAME,
+      constants.SUBMIT_FIELD_TYPENAME
+    ]);
+
+    this.onCategoryChange = this.onCategoryChange.bind(this);
+    this.onExpandCategories = this.onExpandCategories.bind(this);
+    this.onSubmit = this.onSubmit.bind(this);
+    this.onGeometryChange = this.onGeometryChange.bind(this);
+    this.onGeometryStyleChange = this.onGeometryStyleChange.bind(this);
+    this.onCheckboxFieldChange = this.onCheckboxFieldChange.bind(this);
+    this.onAttachmentFieldChange = this.onAttachmentFieldChange.bind(this);
+    this.onFieldChange = this.onFieldChange.bind(this);
+
+    this.props.map.on("dragend", () => {
+      !this.state.isMapPositioned 
+        && this.setState({ isMapPositioned: true });
+    });
+  }
+
+  resetState() {
+    const isSingleCategory = (this.visibleCategories.length === 1);
+
+    this.setState({
+      isFormVisible: false,
+      selectedCategory: isSingleCategory ? this.visibleCategories[0] : null,
+      isCategoryMenuCollapsed: false,
+      isCategoryMenuHidden: isSingleCategory,
+      fieldValues: {},
+      isFormSubmitting: false,
+      formValidationErrors: [],
+      coverImages: [],
+      isMapPositioned: false
+    });
+    this.resetCategoryData();
+  }
+
+  resetCategoryData() {
+    this.geometry = null;
+    this.geometryStyle = null;
+    this.isWithCustomGeometry = false;
+    this.requiredFields = [];
+    this.richTextFields = [];
+    this.richTextImages = [];
+    this.coverImages = [];
+  }
+
+  componentWillMount() {
+    this.visibleCategories = this.props.placeConfig.place_detail
+      .filter(config => config.includeOnForm)
+      .filter(config => {
+        return (
+          !(config.admin_only 
+            && !Util.getAdminStatus(config.dataset, config.admin_groups))
+        );
+      });
+
+    if (this.visibleCategories.length === 1) {
+      this.setActiveCategoryState(this.visibleCategories[0]);
+      this.setState({ isCategoryMenuHidden: true });
+    }
+  }
+
+  componentDidMount() {
+    this.setState({ isFormVisible: true });
+    new Spinner(Shareabouts.smallSpinnerOptions)
+      .spin(document.getElementsByClassName("input-form__submit-spinner")[0]);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.isContinuingFormSession) {
+      this.setActiveCategoryState(
+        this.getCategoryConfig(this.state.selectedCategory)
+      );
+    } else if (nextProps.isLeavingForm) {
+      this.resetState();
+    }
+  }
+
+  getCategoryConfig(category) {
+    return (
+      this.props.placeConfig.place_detail
+        .find(config => config.category === category)
+    );
+  }
+
+  // General handler for field change events.
+  onFieldChange(evt, property) {
+    const nextState = update(
```

Comment:
Why are we using this `update` package? It also has the following description at the top of its man page:

>Note: This is a legacy React addon, and is no longer maintained.
>We don't encourage using it in new code, but it exists for backwards compatibility.

https://www.npmjs.com/package/react-addons-update

Plus, I'm skeptical whether we should be relying on so much mutation. In these situations, I would suggest an approach like immutable-js: https://www.youtube.com/watch?v=I7IdS-PbEgI

Also, I think a "Field" abstraction would help reduce the complexity in `this.state.fieldValies`.

Reply:
I didn't realize this was deprecated. I started using the `update` package because it was recommended in a book, but looks like it's not maintained anymore.

Reply:
But what was the context for choosing to use `update` here? Was it the mutation of a nested object?

Anyway, that pattern may no longer be necessary once we move to a Field abstraction.

Reply:
My understanding is that when updating state it's better not to change nested state in place, but rather create an entirely new object for performance reasons. The `update` package is supposed to help with this--the alternative I think is to use higher-order functions like `concat` that return completely new array references.

But it sounds like immutable-js is designed for the same purpose?

Reply:
>But it sounds like immutable-js is designed for the same purpose?

Yup. That video I linked above sums this issue up pretty well.